Sofa/framework/Type/src/sofa/type/Quat.h

```diff
@@ -143,58 +325,370 @@ class Quat
+
+        Vec3 vEuler;
+        vEuler[0] = atan2(2 * (q[3] * q[0] + q[1] * q[2]), (1 - 2 * (q[0] * q[0] + q[1] * q[1])));   //roll
+        vEuler[1] = asin(y); // pitch
+        vEuler[2] = atan2(2 * (q[3] * q[2] + q[0] * q[1]), (1 - 2 * (q[1] * q[1] + q[2] * q[2])));   //yaw
+        return vEuler;
+    }
 
     /*! Returns the slerp interpolation of Quaternions \p a and \p b, at time \p t.
      \p t should range in [0,1]. Result is \p a when \p t=0 and \p b when \p t=1.
      When \p allowFlip is \c true (default) the slerp interpolation will always use the "shortest path"
      between the Quaternions' orientations, by "flipping" the source Quaternion if needed (see
      negate()). */
-    void slerp(const Quat& a, const Quat& b, Real t, bool allowFlip=true);
+    constexpr void slerp(const Quat& a, const Quat& b, Real t, bool allowFlip=true)
+    {
+        Real cosAngle = (a[0] * b[0] + a[1] * b[1] + a[2] * b[2] + a[3] * b[3]);
+
+        Real c1, c2;
+        // Linear interpolation for close orientations
+        if ((1.0 - std::abs(cosAngle)) < 0.01)
+        {
+            c1 = 1.0f - t;
+            c2 = t;
+        }
+        else
+        {
+            // Spherical interpolation
+            Real angle = acos(std::abs(cosAngle));
+            Real sinAngle = sin(angle);
+            c1 = sin(angle * (1.0f - t)) / sinAngle;
+            c2 = sin(angle * t) / sinAngle;
+        }
+
+        // Use the shortest path
+        if (allowFlip && (cosAngle < 0.0f))
+            c1 = -c1;
+
+        _q[0] = c1 * a[0] + c2 * b[0];
+        _q[1] = c1 * a[1] + c2 * b[1];
+        _q[2] = c1 * a[2] + c2 * b[2];
+        _q[3] = c1 * a[3] + c2 * b[3];
+    }
 
     /// A useful function, builds a rotation matrix in Matrix based on
     /// given quaternion.
-    void buildRotationMatrix(Real m[4][4]) const;
-    void writeOpenGlMatrix( double* m ) const;
-    void writeOpenGlMatrix( float* m ) const;
+    constexpr void buildRotationMatrix(Real m[4][4]) const
+    {
+        m[0][0] = (1 - 2 * (_q[1] * _q[1] + _q[2] * _q[2]));
+        m[0][1] = (2 * (_q[0] * _q[1] - _q[2] * _q[3]));
+        m[0][2] = (2 * (_q[2] * _q[0] + _q[1] * _q[3]));
+        m[0][3] = 0;
+
+        m[1][0] = (2 * (_q[0] * _q[1] + _q[2] * _q[3]));
+        m[1][1] = (1 - 2 * (_q[2] * _q[2] + _q[0] * _q[0]));
+        m[1][2] = (2 * (_q[1] * _q[2] - _q[0] * _q[3]));
+        m[1][3] = 0;
+
+        m[2][0] = (2 * (_q[2] * _q[0] - _q[1] * _q[3]));
+        m[2][1] = (2.0f * (_q[1] * _q[2] + _q[0] * _q[3]));
+        m[2][2] = (1.0f - 2.0f * (_q[1] * _q[1] + _q[0] * _q[0]));
+        m[2][3] = 0;
+
+        m[3][0] = 0;
+        m[3][1] = 0;
+        m[3][2] = 0;
+        m[3][3] = 1;
+    }
+
+    constexpr void writeOpenGlMatrix(double* m ) const
+    {
+        return getOpenGlMatrix<Quat, double>(*this, m);
+    }
+
+    constexpr void writeOpenGlMatrix(float* m) const
+    {
+        return getOpenGlMatrix<Quat, float>(*this, m);
+    }
```

Comment:
I think those should be free function in the opengl module. It has nothing to do in the Quat class

Reply:
True that, but I guess I will do it in a 2nd PR, this one was not breaking until then :p

Reply:
I agree